pkg/restapi/device/operations_test.go

```diff
@@ -302,20 +317,201 @@ func TestRegistration_FinishRegistration(t *testing.T) {
 	})
 }
 
-func TestRegistration_MarshallResponse(t *testing.T) {
-	t.Run("marshall response error ", func(t *testing.T) {
-		value := make(chan int)
-		_, err := json.Marshal(value)
+func TestRegistration_BeginLogin(t *testing.T) {
+	o, err := New(config(t))
+	require.NoError(t, err)
+
+	webAuthConfig := &webauthn.Config{
+		RPDisplayName: "test",
+		RPID:          "test_121",
+	}
+
+	o.webauthn = &webauthn.WebAuthn{
+		Config: webAuthConfig,
+	}
+	userSub := uuid.New().String()
+	o.store.cookies = &cookie.MockStore{
+		Jar: &cookie.MockJar{
+			Cookies: map[interface{}]interface{}{
+				deviceCookieName: userSub,
+			},
+		},
+	}
+	usr := user.User{Sub: userSub}
+	err = o.store.users.Save(&usr)
+	require.NoError(t, err)
+	t.Run("begin login - success", func(t *testing.T) {
+		w := httptest.NewRecorder()
+		cred := webauthn.Credential{}
+		// create device for user
+		deviceInfo := NewDevice(&usr)
+		deviceInfo.AddCredential(cred)
+
+		err := o.saveDeviceInfo(deviceInfo)
+		require.NoError(t, err)
+
+		o.beginLogin(w, newDeviceLoginRequest(userSub))
+		require.Equal(t, http.StatusOK, w.Code)
+	})
+	t.Run("begin login - failed to begin login no Credentials for user", func(t *testing.T) {
+		w := httptest.NewRecorder()
+		deviceInfo := NewDevice(&usr)
+
+		err := o.saveDeviceInfo(deviceInfo)
+		require.NoError(t, err)
+
+		o.beginLogin(w, newDeviceLoginRequest(userSub))
+		require.Equal(t, http.StatusInternalServerError, w.Code)
+	})
+	t.Run("begin login - failed to get user data", func(t *testing.T) {
+		c, err := New(config(t))
+		require.NoError(t, err)
+		userSub := uuid.New().String()
+
+		w := httptest.NewRecorder()
+
+		c.beginLogin(w, newDeviceLoginRequest(userSub))
+		require.Equal(t, http.StatusBadRequest, w.Code)
+	})
+	t.Run("begin login - missing username parameter", func(t *testing.T) {
+		w := httptest.NewRecorder()
+
+		o.beginLogin(w, newDeviceRegistrationRequest())
+		require.Equal(t, http.StatusBadRequest, w.Code)
```

Comment:
You should consider checking for specific errors in your tests. If you change this code later on and your code now erroneously returns top level error all the time all of these tests will still pass. In this case you can catch coding issues only if you inspect code coverage results - unit tests will not catch it.

Reply:
done